src/simplay-web/src/SimulationSpooler.ts

```diff
@@ -23,36 +26,88 @@ export class SimulationSpooler {
     createEntities(this.context);
   }
 
-  run(speedFactor = 1) {
-    throw Error('TODO implement');
+  private spoolTimestamp(timestamp: number) {
+    const events = this.simulationData.events.filter(
+      (event) => Math.round(event.timestamp) === timestamp
+    );
+    events.forEach((event) => {
+      event.execute(this.context);
+    });
   }
 
-  pause() {
-    throw Error('TODO implement');
+  async run() {
+    const maxTimestamp = Math.max(
+      ...this.simulationData.events.map((event) => event.timestamp)
+    );
+    while (!this.stopRequested) {
+      const frameDuration = 1000 / this.speedFactor;
+      const now = Date.now();
+
+      this.spoolTimestamp(this.currentSimTimeStamp);
+
+      const executionDuration = now - Date.now();
+      if (this.currentSimTimeStamp > maxTimestamp) {
+        this.stopRequested = true;
+      }
+      await new Promise((resolve) =>
+        setTimeout(resolve, frameDuration - executionDuration)
+      );
+      this.currentSimTimeStamp++;
+    }
+    this.stopRequested = false;
+  }
+
+  async pause() {
+    this.stopRequested = true;
+    // Wait for the current step to finish
+    await new Promise((resolve) =>
+      setTimeout(resolve, 1000 / this.speedFactor)
+    );
   }
 
-  continue() {
```

Comment:
This is no longer needed, because pause can be called instead to continue with the spooling, right?